### PR TITLE
Support rendering recursive static DAGs

### PR DESCRIPTION
**Pull request by rileyjbauer**
With this change, `StaticGraphParser` no longer breaks when trying to lay out recursive pipelines.

This PR also adds a new mock template: `mock-recursive-template.yaml` which includes multiple instances of recursion.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/845)

### CODE REVIEW

File: frontend/src/lib/StaticGraphParser.ts [hunk 1]
```diff
@@ -76,72 +76,89 @@ export function _populateInfoFromTemplate(info: SelectedNodeInfo, template?: Tem
...
-            buildDag(graph, task.template, templates, nodeId);
-          } else if (child.nodeType === 'container' ) {
-            _populateInfoFromTemplate(info, child.template);
-          } else {
-            logger.error(`Unknown nodetype: ${child.nodeType} on template: ${child.template}`);
-          }
-        }
+      // Parent here will be the task that pointed to this DAG template.
+      // We do not set an edge if the task has dependencies because in that case it makes sense
+      // for incoming edges to only be drawn from its dependencies, not the DAG node itself
```
Comment (yebrahim) [inline frontend/src/lib/StaticGraphParser.ts]: Can you clarify this comment a little? Why does having or not having dependencies affect having an edge to its parent node?

File: frontend/src/lib/StaticGraphParser.ts [hunk 2]
```diff
@@ -76,72 +76,89 @@ export function _populateInfoFromTemplate(info: SelectedNodeInfo, template?: Tem
...
+      // "Child" here is the template that this task points to. This template should either be a
+      // DAG, in which case we recurse, or a container which can be thought of as a leaf node
+      const child = templates.get(task.template);
+      if (child) {
+        if (child.nodeType === 'dag') {
+          buildDag(graph, task.template, templates, alreadyVisited, nodeId);
+        } else if (child.nodeType === 'container' ) {
+          _populateInfoFromTemplate(info, child.template);
+        } else {
+          logger.error(`Unknown nodetype: ${child.nodeType} on template: ${child.template}`);
```
Comment (yebrahim) [inline frontend/src/lib/StaticGraphParser.ts]: Should this error be shown? This will have been a user error (uploading a pipeline with incorrect format), and we should probably surface that.

File: frontend/src/lib/StaticGraphParser.ts (hunk 1 above)
Comment (rileyjbauer) [reply frontend/src/lib/StaticGraphParser.ts]: This might be a bit wordy, see if you think it makes sense:
```
// Parent here will be the task that pointed to this DAG template.
// Within a DAG template, tasks can have dependencies on one another, and long chains of
// dependencies can be present within a single DAG. In these cases, we choose not to draw an
// edge from the DAG node itself to these tasks with dependencies because such an edge would
// not be meaningful to the user. For example, consider a DAG A with two tasks, B and C, where
// task C depends on the output of task B. C is a task of A, but it's much more important that
// C depends on B, so to avoid cluttering the graph, we simply omit the edge between A and C:
//      A                  A   
//    /   \    becomes    /  
//   B <-- C             B 
//                      /
//                     C

File: frontend/src/lib/StaticGraphParser.ts (hunk 2 above)
Comment (rileyjbauer) [reply frontend/src/lib/StaticGraphParser.ts]: Hm, I think you're right.

File: frontend/src/lib/StaticGraphParser.ts (hunk 1 above)
Comment (yebrahim) [reply frontend/src/lib/StaticGraphParser.ts]: Are we sure we won't ever be removing useful information from the graph by doing so?
Separately, the comment is much more clear now, thanks.

Comment (rileyjbauer) [reply frontend/src/lib/StaticGraphParser.ts]: I don't know for certain, but I haven't been able to think of a case where it'd be preferable to see those extra edges, and removing them makes the execution flow of the pipeline a lot clearer (IMO)

Comment (yebrahim) [reply frontend/src/lib/StaticGraphParser.ts]: SGTM then, thanks.
/lgtm
